internal/ebpf/probesfacade/captor/capture.go

```diff
@@ -107,51 +119,110 @@ func Capture(functionSymbol string, cmdArgs []string, opts CaptureOptions) ([]ui
 	if err != nil {
 		return nil, fmt.Errorf("error initializing map (%s) with PerfBuffer: %v", bpfEventsMap, err)
 	}
 
-	// run args that we want to trace
+	return &ebpfSetup{
+		mod:      bpfModule,
+		link:     traceLink,
+		pb:       pb,
+		eventsCh: eventsChannel,
+		lostCh:   lostChannel,
+		opts:     opts,
+		cmd:      cmdArgs,
+	}, nil
+}
+
+// Close closes the ebpf link and module.
+func (ebpf *ebpfSetup) Close() {
+	ebpf.link.Destroy()
+	ebpf.mod.Close()
+}
+
+// Capture collects syscalls from the executed command.
+// Returns values through the given channels.
+func (ebpf *ebpfSetup) Capture(collectedSyscalls chan []uint32, errorCh chan error) {
+	// setting up ticker to dump results
+	// every interval of time.
+	var ticker *time.Ticker
+	var interval time.Duration
+	if ebpf.opts.Interval > 0 {
+		interval = time.Duration(ebpf.opts.Interval) * time.Second
+	} else {
+		// wait what? ~114 years of interval?
+		// yes, so that the ticker will never be triggered.
+		// the value looks reasonably good.
+		interval = time.Duration(1000000) * time.Hour
+	}
+	ticker = time.NewTicker(interval)
+	defer ticker.Stop()
+
 	var wg sync.WaitGroup
 	wg.Add(1)
-	outputCh := make(chan string)
-	errorCh := make(chan string)
-	go executor.Run(cmdArgs, opts.CommandOutput, opts.CommandError, &wg, outputCh, errorCh)
+	cmdStdoutCh := make(chan string)
+	cmdStderrCh := make(chan string)
+	// running command to trace its syscalls
+	go executor.Run(
+		ebpf.cmd,
+		ebpf.opts.CommandOutput,
+		ebpf.opts.CommandError,
+		&wg,
+		cmdStdoutCh,
+		cmdStderrCh,
+	)
 
 	var syscalls []uint32
 	go func() {
 		for {
 			select {
-			case data := <-eventsChannel:
+			case data := <-ebpf.eventsCh:
 				var e event
 				err := binary.Read(bytes.NewBuffer(data), binary.LittleEndian, &e)
 				if err != nil {
-					fmt.Fprintf(os.Stderr, "error reading event: %v\n", err)
+					//fmt.Fprintf(os.Stderr, "error reading event: %v\n", err)
+					// will be left empty for now.
 					return
 				}
 				syscalls = append(syscalls, e.SyscallID)
-			case lost := <-lostChannel:
-				fmt.Fprintf(os.Stderr, "lost %d data\n", lost)
+			case <-ticker.C:
+				// used to send incremental result
+				// every interval of time.
+				collectedSyscalls <- syscalls
+				// we clear the slice after sending data
+				// to the channel, so on the next iteration
+				// this will have only the most recent values.
+				syscalls = nil
+			case _ = <-ebpf.lostCh:
+				// managing errors from libbpf
+				// will be left empty for now.
+				//fmt.Fprintf(os.Stderr, "lost %d data\n", lost)
 				return
-			case line, ok := <-outputCh:
+			case line, ok := <-cmdStdoutCh:
+				// managing stdout from executed command
 				if !ok {
 					break
 				}
-				fmt.Println(line)
-			case err, ok := <-errorCh:
+				fmt.Println("stdout:", line)
+			case err, ok := <-cmdStderrCh:
+				// managing stderr from executed command
 				if !ok {
 					break
 				}
-				fmt.Println(err)
+				fmt.Println("stderr:", err)
```

Comment:
we should send the error to the errorCh (the one in input) so that the above context can correctly handle the command error and exit accordingly

Reply:
keep in mind that this is the stderr generated from the executed command. it could happen that the command generate some error without failing. IMO we should not send it through the `errorCh`, since this will stop the execution of the command.

Reply:
For the future, but in that case I would return specific errors and filter at consumer side for ones that should make it stop or not

Reply:
If you want to keep it fine, it will be handy for the future